File: dashboard/lib/fusion.ts
/**
 * Signal fusion helpers — match scanner/fusion.py output.
 *
 * Scanner writes `meta.fusion_group_id` / `meta.fusion_primary` /
 * `meta.confidence_score` / `meta.fused_sources` on each signal row.
 * 这里只是按 fusion_group_id 重组成 SignalCard 友好的 (primary, related[])。
 *
 * 老数据（V2.9 之前）没有 fusion_group_id —— 自动回退为单信号即组。
 */
import type { Signal } from "@/lib/types";

export interface FusionGroup {
  primary: Signal;
  related: Signal[];          // 同组其它非 primary 信号
  confidence: number;         // 1-5
  fused_sources: string[];    // 同组所有不同 source 名
}

function groupKey(s: Signal): string {
  const gid = s.meta?.fusion_group_id;
  if (typeof gid === "string" && gid.length > 0) return gid;
  // 回退：用 inst_id + detected_at 分钟桶（5min）保证老数据也能分组
  const t = new Date(s.detected_at).getTime();
  const bucket = Math.floor(t / 60_000 / 5);
  return `${s.inst_id}|${bucket * 5}`;
}

export function fuseSignals(signals: Signal[]): FusionGroup[] {
  const buckets = new Map<string, Signal[]>();
  for (const s of signals) {
    const k = groupKey(s);
    const arr = buckets.get(k) ?? [];
    arr.push(s);
    buckets.set(k, arr);
  }

  const groups: FusionGroup[] = [];
  for (const arr of buckets.values()) {
    // primary = 标记位有就用，否则取 |chg_pct| 最大
    const primary =
      arr.find((s) => s.meta?.fusion_primary === true) ??
      arr.slice().sort((a, b) => Math.abs(Number(b.chg_pct)) - Math.abs(Number(a.chg_pct)))[0];
    const related = arr.filter((s) => s.id !== primary.id);
    const distinctSources = new Set(arr.map((s) => s.source));
    const confidence =
      (primary.meta?.confidence_score as number | undefined) ??
      Math.min(5, distinctSources.size);
    const fusedSources =
      (primary.meta?.fused_sources as string[] | undefined) ?? Array.from(distinctSources);
    groups.push({
      primary,
      related,
      confidence,
      fused_sources: fusedSources,
    });
  }

  // 排序：confidence 高优先，其次 primary.detected_at 新优先
  groups.sort((a, b) => {
    if (b.confidence !== a.confidence) return b.confidence - a.confidence;
    return (
      new Date(b.primary.detected_at).getTime() -
      new Date(a.primary.detected_at).getTime()
    );
  });
  return groups;
}

File: scanner/fusion.py
"""Signal fusion — 多 monitor 同币种共振检测（V2.9 / Phase 3）。

底层逻辑：同一币种在同一轮扫描里被多个 monitor 同时命中，比单一维度信号
强得多。融合层不丢弃任何 raw signal（backtest 要用），而是给每个信号打
上 fusion 元数据：

  meta.confidence_score: 1-5（同 inst_id 命中的不同 source 数，capped）
  meta.fused_sources:    list[str]（同组所有 source 名）
  meta.fusion_primary:   bool（同组里 |chg_pct| 最大的标记为 primary）
  meta.fusion_group_id:  str（inst_id|time_bucket，便于 dashboard JOIN）

下游：
- FeishuNotifier 渲染时按 primary 折叠展示，related 信号变成"+N 维度"chip
- Dashboard 用 fusion_group_id 服务端 group，SignalCard.related[] 自动填
- Supabase 仍保留每行 raw signal（meta 里带 fusion 标记），backtest 不受影响
"""
from collections import defaultdict


def fuse(signals, time_bucket_min: int = 5):
    """对一组 fresh_signals 做就地融合标记（不改变 list 长度，只丰富 meta）。

    time_bucket_min: 同 inst_id 信号在多少分钟内算同一融合组。默认 5min。
    Scanner 单轮扫描通常 1-3 分钟跑完，所以 5min 足够把同轮所有同币信号归一组。
    """
    if not signals:
        return signals

    # 按 inst_id + time bucket 分组
    groups = defaultdict(list)
    for s in signals:
        bucket = (s.bar_ts_ms // 1000 // 60 // time_bucket_min) if s.bar_ts_ms else 0
        groups[(s.inst_id, bucket)].append(s)

    for (inst_id, bucket), group in groups.items():
        distinct_sources = list({s.source for s in group})
        confidence = min(5, len(distinct_sources))
        # primary = |chg_pct| 最大的（最显著的那条作为代表）
        primary = max(group, key=lambda s: abs(s.chg_pct))
        group_id = f"{inst_id}|{bucket * time_bucket_min}"

        for s in group:
            # 不覆盖现有 meta 字段，只增量加 fusion 标记
            if not isinstance(s.meta, dict):
                s.meta = {}
            s.meta["confidence_score"] = confidence
            s.meta["fused_sources"] = distinct_sources
            s.meta["fusion_primary"] = (s is primary)
            s.meta["fusion_group_id"] = group_id

    return signals


def primary_signals(signals):
    """从已 fuse 的 signals 列表里抽出 primary（每组一条）。Feishu 推送时用。"""
    return [s for s in signals if s.meta.get("fusion_primary", True)]
